Userland/entr.cpp

```diff
@@ -0,0 +1,203 @@
+        .short_name = 's',
+        .accept_value = [&](auto) {
+            executable_name = shell;
+            arguments.prepend("-c");
+            return true;
+        },
+    });
+    args_parser.add_positional_argument(arguments, "Command to execute", "command");
+    args_parser.parse(argc, argv);
+
+    if (executable_name == nullptr)
+        executable_name = arguments.take_first();
+
+    Vector<String> str_arguments;
+    str_arguments.ensure_capacity(arguments.size());
+    for (auto& arg : arguments)
+        str_arguments.unchecked_append(arg);
+
+    Core::EventLoop loop;
+
+    auto file = Core::File::stdin();
+    for (;;) {
+        auto line = file->read_line(PATH_MAX);
+        if (line.is_null())
+            break;
+
+        auto line_view = StringView { line.data(), line.size() };
+        String filename = line_view.trim_whitespace(TrimMode::Right);
+        auto real_path = Core::File::real_path_for(filename);
+
+        if (verbose)
+            warnln("[path] {} is {}", filename, real_path);
+
+        auto watch_fd = watch_file(real_path.characters(), real_path.length());
+        if (watch_fd < 0) {
+            perror("watch_file");
+            return 0;
+        }
+        fcntl(watch_fd, F_SETFD, FD_CLOEXEC);
+        auto notifier = Core::Notifier::construct(watch_fd, Core::Notifier::Event::Read);
+        if (verbose)
+            warnln("[watch_file] watching {} with fd={}", filename, watch_fd);
+
+        notifier->on_ready_to_read = [notifier, verbose, clear, force_restart, path = real_path, &executable_name, str_arguments = prepare_arguments(arguments, filename, real_path), m_old_pid = Optional<pid_t> {}] {
+            auto& mutable_old_pid = const_cast<Optional<pid_t>&>(m_old_pid);
+            if (clear)
+                out("\x1b[H\x1b[2J\x1b[3J");
+            char buffer[32];
+            int rc = read(notifier->fd(), buffer, sizeof(buffer));
+            ASSERT(rc >= 0);
+
+            if (verbose)
+                warnln("[notify] Detected change in {}", path);
+
+            if (verbose) {
+                StringBuilder builder;
+                builder.join(' ', str_arguments);
+                warnln("[exec] exec {} with ({})", executable_name, builder.string_view());
+            }
+
+            if (m_old_pid.has_value()) {
+                if (force_restart) {
+                    if (verbose)
+                        warnln("[restart] Killing previous instance for {} with pid={}", path, m_old_pid.value());
+                    kill(m_old_pid.value(), SIGTERM);
+                }
+                int wstatus = 0;
+                waitpid(m_old_pid.value(), &wstatus, 0);
+                mutable_old_pid.clear();
+            }
+
+            Vector<const char*> argv;
+            argv.append(executable_name);
+            for (auto arg : str_arguments)
+                argv.append(arg.characters());
+            argv.append(nullptr);
+
+            posix_spawn_file_actions_t action;
```

Comment:
Doesn't look like you do anything with this other than create and destroy it -- can you just pass nullptr for the action instead?

Reply:
can I? I'm not too familiar with `posix_spawn()`, its manpage always talks about "the object pointed to by `attrp`" so I don't think it can be null?

Reply:
Both Serenity's man page and https://pubs.opengroup.org/onlinepubs/009695399/functions/posix_spawn.html are fairly explicit that it can be nullptr.

Reply:
I is bad at reading manpages, apparently.